File: abctpy/abct/moderemoval.py
from typing import Literal
from numpy.typing import ArrayLike
from pydantic import validate_call, ConfigDict

import numpy as np
from scipy import sparse


@validate_call(config=ConfigDict(arbitrary_types_allowed=True))
def moderemoval(
    X: ArrayLike, type: Literal["degree", "global", "rankone", "soft"] = "degree"
) -> np.ndarray:

    X = np.asarray(X)

    match type:
        # Degree correction
        case "degree":
            if not np.all(X >= 0):
                raise ValueError("Invalid degree correction: Matrix must be non-negative.")
            So = np.sum(X, axis=1)
            Si = np.sum(X, axis=0)
            return X - np.outer(So, Si) / np.sum(So)

        # Global signal regression
        case "global":
            G = np.mean(X, axis=0, keepdims=True)
            return X - (X @ G.T) @ G / (G @ G.T)

        # Subtraction of rank-one approximation
        case "rankone":
            U, S, VT = sparse.linalg.svds(X, k=1)
            return X - U @ np.diag(S) @ VT

        # Soft removal of primary modes
        case "soft":
            if not (X.shape[0] == X.shape[1] and np.allclose(X, X.T)):
                raise ValueError("Invalid soft mode removal: Network matrix must be symmetric.")

            # Compute eigendecomposition
            D, V = np.linalg.eigh(X)
            idx = np.argsort(D)[::-1]
            V = V[:, idx]
            D = D[idx]

            n = len(D)

            # Fit cubic polynomial to all eigenvalues
            bk, r0 = np.polyfit(np.arange(1, n + 1), D, 3, full=True)[:2]
            rms0 = np.sqrt(r0[0] / n)  # get rms

            # Rescale x to [0, 1]
            x = (np.arange(1, n + 1) - 1) / (n - 1)
            y = np.zeros(n)

            # Find optimal fit
            for k in range(n):
                b = bk
                bk, rk = np.polyfit(np.arange(k + 1, n + 1), D[k:], 3, full=True)[:2]
                assert len(np.arange(k + 1, n + 1)) == (n - k)
                rmsk = np.sqrt(rk[0] / (n - k))
                y[k] = (rms0 - rmsk) / rms0
                # Detect knee of optimal fit and break
                if k > 0 and (y[k] - x[k]) < (y[k - 1] - x[k - 1]):
                    break

            # Apply the optimal fit
            D = np.polyval(b, np.arange(1, n + 1))
            return V @ np.diag(D) @ V.T


with open("abct/docs/moderemoval", "r") as f:
    moderemoval.__doc__ = f.read()

File: abct-matlab/abct/moderemoval.m
function X = moderemoval(X, type)
% MODEREMOVAL Mode removal from network or timeseries data
%
%   X1 = moderemoval(X)
%   X1 = moderemoval(X, type)
%
%   Inputs:
%       X:  Network matrix of size n x n, or data matrix of size n x p.
%           n is the number of nodes or data points and
%           p is the number of features.
%
%       type: Type of mode removal.
%           "degree": Degree correction (default).
%           "global": Global signal regression.
%           "rankone": Subtraction of rank-one approximation.
%           "soft": Soft removal of primary modes.
%
%   Outputs:
%       X1: Network or timeseries matrix after mode removal.
%
%   Methodological notes:
%       Degree correction, global signal regression, and subtraction of
%       rank-one approximation all produce approximatley equivalent
%       results. The "soft" mode removal makes the network sparse by using
%       cubic interpolation to "despike" an initial eigenspectrum peak.
%
%   See also:
%       DEGREES, GRADIENTS.

arguments
    X (:, :) double {mustBeNonempty, mustBeFinite, mustBeReal}
    type (1, 1) string {mustBeMember(type, ["degree", "global", "rankone", "soft"])} = "degree"
end

switch type
    % Degree correction
    case "degree"
        assert(all(X >= 0, "all"), ...
            "Invalid degree correction: Matrix must be non-negative.")
        So = sum(X, 2);
        Si = sum(X, 1);
        X = X - So * Si / sum(So);

    % Global signal regression
    case "global"
        G = mean(X, 1);
        X = X - (X * G') * G / (G * G');
        % (X - (X * G') * G / (G * G')) * G' = 0    % verification

    % Subtraction of rank-one approximation
    case "rankone"
        [U, S, V] = svds(X, 1);
        X = X - U * S * V';

    % Soft removal of primary modes
    case "soft"
        assert(isequal(size(X, 1), size(X, 2)) && all(X - X' < eps("single"), "all"), ...
            "Invalid soft mode removal: Network matrix must be symmetric.")
        [V, D] = eig(X, "vector");
        [~, ix] = sort(D, "descend");
        V = V(:, ix);
        D = D(ix);

        n = length(D);
        [bk, r0] = polyfit((1:n)', D, 3);
        rms0 = r0.normr / sqrt(n);    % get rms
        x = rescale(1:n).';
        y(n) = 0;
        for k = 1:n
            b = bk;
            [bk, rk] = polyfit((k:n).', D(k:n), 3);
            assert(numel((k:n))==(n-k+1))
            rmsk = rk.normr / sqrt(n-k+1);
            y(k) = (rms0 - rmsk) / rms0;
            % detect knee of optimal fit and break
            if k > 1 && ((y(k) - x(k)) < (y(k-1) - x(k-1)))
                break
            end
        end

        D = polyval(b, (1:n).');
        X = V * diag(D) * V';
end
